worlds/tunic/er_scripts.py

```diff
@@ -240,9 +243,78 @@ def pair_portals(world: "TunicWorld") -> Dict[Portal, Portal]:
-        portal_pairs, dependent_regions, dead_ends, two_plus = \
-            create_plando_connections(plando_connections, dependent_regions, dead_ends, two_plus)
+        for connection in plando_connections:
+            p_entrance = connection.entrance
+            p_exit = connection.exit
+
+            if p_entrance.startswith("Shop"):
+                p_entrance = p_exit
+                p_exit = "Shop Portal"
+
+            portal1 = None
+            portal2 = None
+
+            # search two_plus for both at once
+            for portal in two_plus:
+                if p_entrance == portal.name:
+                    portal1 = portal
+                if p_exit == portal.name:
+                    portal2 = portal
+
+            # search dead_ends individually since we can't really remove items from two_plus during the loop
+            if not portal1:
+                for portal in dead_ends:
+                    if p_entrance == portal.name:
+                        portal1 = portal
+                        break
+                if not portal1:
+                    raise Exception(f"Could not find entrance named {p_entrance} for "
+                                    f"plando connections in {player_name}'s YAML.")
+                dead_ends.remove(portal1)
+            else:
+                two_plus.remove(portal1)
+
+            if not portal2:
+                for portal in dead_ends:
+                    if p_exit == portal.name:
+                        portal2 = portal
+                        break
+                if p_exit in ["Shop Portal", "Shop"]:
+                    portal2 = Portal(name="Shop Portal", region=f"Shop",
+                                     destination="Previous Region_")
+                    shop_count -= 1
+                    if shop_count < 0:
+                        shop_count += 2
+                    for p in portal_mapping:
+                        if p.name == p_entrance:
+                            shop_scenes.add(p.scene())
+                            break
+                else:
+                    if not portal2:
+                        raise Exception(f"Could not find entrance named {p_exit} for "
+                                        f"plando connections in {player_name}'s YAML.")
+                    dead_ends.remove(portal2)
+            else:
+                two_plus.remove(portal2)
+
+            portal_pairs[portal1] = portal2
+
+            # update dependent regions based on the plando'd connections, to ensure the portals connect well, logically
+            for origins, destinations in dependent_regions.items():
+                if portal1.region in origins:
+                    if portal2.region in non_dead_end_regions:
+                        destinations.append(portal2.region)
+                if portal2.region in origins:
+                    if portal1.region in non_dead_end_regions:
+                        destinations.append(portal1.region)
```

Comment:
This section is to update the Dependent Regions, which are a thing I put together to basically tell the ER pairing scripts which regions you have access to provided you have a starting region. Basically, it assumes you have all items already.

This is updating that dependent regions thing with new regions you can get access to given a certain region.

For example: Let's say you plando the Forest Belltower upper entrance to the West Garden laurels exit. The ER pairing script will now get to know that if you have access to West Garden, you have access to all of Forest Belltower, and vice versa.

If it's a dead end region, we don't want the ER script to care about that being added -- the dead end regions are not relevant for the initial pairing to ensure all non-dead end regions are accessible, so adding any of those in would make it get to its goal region count too early.

Reply:
So the old code was essentially buggy for including these at all, but it was a harmless bug that we're drive-by fixing here?

Reply:
Technically, yeah the plando connections function that I had written was buggy, but it served its purpose at the time -- letting me make Universal Tracker work nicely. It would correctly plando all of the connections (since it would just do a full plando every time). Just when doing the actual plando connections stuff, the issue was noticed.